cli/flox-activations/src/activations.rs

```diff
@@ -1,25 +1,93 @@
 use std::path::{Path, PathBuf};
 
-#[cfg(target_os = "linux")]
-use flox_core::canonical_path::CanonicalPath;
 use flox_core::{path_hash, traceable_path, Version};
 use fslock::LockFile;
+use nix::errno::Errno;
+use nix::sys::signal::kill;
+use nix::unistd::Pid;
 use serde::{Deserialize, Serialize};
-use time::OffsetDateTime;
+use serde_json::json;
+use time::{Duration, OffsetDateTime};
 use tracing::debug;
 use uuid::Uuid;
 
 type Error = anyhow::Error;
 
 /// Deserialized contents of activations.json
-#[derive(Clone, Debug, Deserialize, Serialize)]
-struct Activations {
+///
+/// This is the state of the activations of the environments.
+/// There is EXACTLY ONE [Activations] file per environment path, i.e. .flox directory,
+/// which may be rendered at different times with different store paths.
+/// [Activations::activations] is a list of [Activation]s
+/// with AT MOST ONE activation for a given store path.
+/// This latter invariant is enforced by [Activations::get_or_create_activation_for_store_path]
+/// being the only way to add an activation.
+/// Activations are identifiable by their [Activation::id], for simpler lookups
+/// and global uniqueness in case the that two environments have the same store path.
+///
+/// Notably, the [Activations] does not feature methods to remove activations.
+/// Removing actiavtions falls onto the watchdog, which is responsible for cleaning up activations.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct Activations {
     version: Version<1>,
     activations: Vec<Activation>,
 }
 
+impl Activations {
+    /// Get a mutable reference to the activation with the given ID.
+    ///
+    /// Used internally to manipulate the state of an activation.
+    pub fn activation_for_id_mut(&mut self, activation_id: Uuid) -> Option<&mut Activation> {
+        self.activations
+            .iter_mut()
+            .find(|activation| activation.id == activation_id)
+    }
+
+    /// Get an immutable reference to the activation with the given ID.
+    ///
+    /// Used internally to manipulate the state of an activation.
+    pub fn activation_for_id_ref(&mut self, activation_id: Uuid) -> Option<&Activation> {
+        self.activations
+            .iter()
+            .find(|activation| activation.id == activation_id)
+    }
+
+    /// Get a mutable reference to the activation with the given store path.
+    ///
+    /// If the activation does not exist, create a new one.
+    /// [NewOrExisting] signals whether a new activation was created or an existing one was found.
+    pub fn get_or_create_activation_for_store_path(&mut self, store_path: &Path) -> NewOrExisting {
```

Comment:
blocking: start needs to add PID

nit: I'm not sure this method should be factored out. Or at least I don't think it should create a new `Activation`. I think `start-or-attach` should be the only thing that ever needs it. And moving the logic away from `start-or-attach` seems to make it harder to see what's being done. I think that makes it harder to spot e.g. that we're not adding PID for start

Reply:
Looks like you just went with the double lookup?